log (some) traces directly, avoiding a Feed

tbg opened the pull request:
logging to the Feed can panic at high volumes, and
for now having a callback directly deal with the
traces is more adequate for current use (debugging).

functionally this is a no-op, except that slow
transactions are now always logged, even without
V(2).

tamird commented:
Why use feeds elsewhere? Also, where is the panic?

tschottdorf commented:
an overflowing Feed will panic. Traces may eventually end up in a Feed when they're being written to the database, but we'll want to filter before (callback-style). I suspect that a bunch of places would be happier with a synchronous callback interface than with a Feed (for instance, `Call{Error,Success}Event` which don't do much but increment a counter). If you don't know that the subscriber can handle events in real-time, using a Feed makes sense but then you need to start worrying about the Feed filling up, which also isn't ideal.

tbg committed:
log (some) traces directly, avoiding a Feed

logging to the Feed can panic at high volumes, and
for now having a callback directly deal with the
traces is more adequate for current use (debugging).

functionally this is a no-op, except that slow
transactions are now always logged, even without
V(2).

tbg committed:
add Trace entry around llMu lock

mrtracy commented:
Did you actually observe panicking Feeds in tests, or are you just anticipating them and rejecting them based on the mere possibility of them?  I don't know what good feeds are going to be if we never use them because they can panic in high-volume situations.

spencerkimball commented:
As discussed at lunch, might be better to move the feeds to not being
buffered and therefore not having to have them panic. This seems analogous
to using a synchronous callback.

On Mon, Jul 13, 2015 at 12:04 PM, mrtracy <notifications@github.com> wrote:

> Did you actually observe panicking Feeds in tests, or are you just
> anticipating them and rejecting them based on the mere possibility of them?
> I don't know what good feeds are going to be if we never use them because
> they can panic in high-volume situations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1658#issuecomment-120980693>
> .
>

tschottdorf commented:
yes, this happened in practice on the Put test, though under circumstances in which the system was falling apart anyways. Is this good to go? The feed discussion seems reasonable but out of scope.

tschottdorf commented:
ping

tamird commented:
@tschottdorf I think what @spencerkimball was trying to say is that we decided at lunch that this isn't the right direction to go: if we're allowing the system to accept load until it reaches this panic in the feed then we should implement flow control to make sure that doesn't happen.

spencerkimball commented:
Or maybe the simpler solution is to just change the feeds to be more
synchronous and push back with their own flow control by making them
unbuffered.

On Tue, Jul 14, 2015 at 11:59 AM, Tamir Duberstein <notifications@github.com
> wrote:

> @tschottdorf <https://github.com/tschottdorf> I think what @spencerkimball
> <https://github.com/spencerkimball> was trying to say is that we decided
> at lunch that this isn't the right direction to go: if we're allowing the
> system to accept load until it reaches this panic in the feed then we
> should implement flow control to make sure that doesn't happen.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1658#issuecomment-121289583>
> .
>

mrtracy commented:
Yes, instead of this pull request we could convert all of our current feeds into synchronous broadcast events.  The current feeds could be converted to synchronous broadcast with two changes:

1. remove the select statement in the feed's Publish method
2. make the subscriber channels unbuffered.

Of course, there would be more work to do in terms of comments and renaming. It's really not a feed at that point, really just a publisher.

tschottdorf commented:
This isn't a direction, this is KISS. Previous code sends everything through a feed into an object that @mrtracy tells me isn't the right recipient. The actual processing for traces (sampling, in-memory buffers and persisting a subset to the KV) hasn't been built yet. So really the only decision about this PR is whether to keep that unecessary complexity or build it as we actually build it.
There's been so much discussion about what I think is a trivial PR that has little to do with the Feed system - weird that it's so easy to put in a Feed but so hard to take one out. In any case, I've since identified the problem that actually overflowed the feeds, and I'll just fix that and close this PR. I'm going to open an issue for reviewing the Feed system.
